types/react-query/index.d.ts

```diff
@@ -1,142 +1,370 @@
+export type AnyQueryKey =
+    | readonly [QueryKeyPart]
+    | readonly [QueryKeyPart, QueryKeyPart]
+    | readonly [QueryKeyPart, QueryKeyPart, QueryKeyPart]
+    | readonly [QueryKeyPart, QueryKeyPart, QueryKeyPart, QueryKeyPart];
+export type AnyVariables =
+    | readonly []
+    | readonly [any]
+    | readonly [any, any]
+    | readonly [any, any, any]
+    | readonly [any, any, any, any]
+    | readonly [any, any, any, any, any];
+
+export type QueryFunction<TResult, TKey extends AnyQueryKey> = (...key: TKey) => Promise<TResult>;
+export type QueryFunctionWithVariables<TResult, TKey extends AnyQueryKey, TVariables extends AnyVariables> = (
+    ...key: _.List.Concat<TKey, TVariables>
+) => Promise<TResult>;
+
+export type InfiniteQueryFunction<TResult, TKey extends AnyQueryKey, TMoreVariable> = (
+    ...keysAndMore: _.List.Append<TKey, TMoreVariable> | TKey
+) => Promise<TResult>;
+export type InfiniteQueryFunctionWithVariables<
+    TResult,
+    TKey extends AnyQueryKey,
+    TVariables extends AnyVariables,
+    TMoreVariable
+> = (
+    ...keysAndVariablesAndMore:
+        | _.List.Append<_.List.Concat<TKey, TVariables>, TMoreVariable>
+        | _.List.Concat<TKey, TVariables>
+) => Promise<TResult>;
+
+export interface BaseQueryOptions {
+    /**
+     * Set this to `true` to disable automatic refetching when the query mounts or changes query keys.
+     * To refetch the query, use the `refetch` method returned from the `useQuery` instance.
+     */
     manual?: boolean;
+    /**
+     * If `false`, failed queries will not retry by default.
+     * If `true`, failed queries will retry infinitely.
+     * If set to an integer number, e.g. 3, failed queries will retry until the failed query count meets that number.
+     */
     retry?: boolean | number;
     retryDelay?: (retryAttempt: number) => number;
     staleTime?: number;
     cacheTime?: number;
     refetchInterval?: false | number;
     refetchIntervalInBackground?: boolean;
     refetchOnWindowFocus?: boolean;
-    onError?: (err: any) => void;
-    onSuccess?: (data: TResult) => void;
+    refetchOnMount?: boolean;
+    onError?: (err: unknown) => void;
     suspense?: boolean;
+}
+
+export interface QueryOptions<TResult> extends BaseQueryOptions {
+    onSuccess?: (data: TResult) => void;
+    onSettled?: (data: TResult | undefined, error: unknown | null) => void;
     initialData?: TResult;
 }
 
-export interface QueryOptionsPaginated<TResult> extends QueryOptions<TResult> {
-    paginated: true;
-    getCanFetchMore: (lastPage: TResult, allPages: TResult[]) => boolean;
+export interface InfiniteQueryOptions<TResult, TMoreVariable> extends QueryOptions<TResult> {
+    getFetchMore: (lastPage: TResult, allPages: TResult[]) => TMoreVariable | false;
 }
 
-export interface QueryResult<TResult, TVariables> {
-    data: null | TResult;
-    error: null | Error;
-    isLoading: boolean;
+export interface QueryResultBase<TResult> {
```

Comment:
Although you're right that we can't conclude `data` based on `success` status, discriminated union here still make sense for `error` and `loading` statuses.

Reply:
So, unfortunately, we cannot do this for `error` either. This is because the type of error is not `Error` (built-in class) but any possible throwable type (the library does not wrap inner exceptions), which can be anything, including primitives, `null` or `undefined`. That makes the type be `unknown | null` which is just `unknown`. So any CFA based on that does not make sense.

Reply:
To clarify, `const { status, error } = useQuery(key, () => throw null)` would produce `status === 'error' && error === null`.

Reply:
Added discriminated union that allows to have `status === 'success' && data !== TResult`